src/Algebra/Bundles/Free/Magma.agda

```diff
@@ -0,0 +1,262 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Definitions of 'pre-free' and 'free' magma
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Bundles.Free.Magma where
+
+open import Algebra.Core
+import Algebra.Definitions as Defs using (Congruent₂)
+import Algebra.Structures as Strs using (IsMagma)
+open import Algebra.Morphism.Structures using (IsMagmaHomomorphism)
+open import Algebra.Bundles using (Magma)
+open import Algebra.Bundles.Raw using (RawMagma)
+open import Function.Base using (id; _∘_)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Binary.Morphism using (IsRelHomomorphism)
+open import Level using (Level; _⊔_)
+open import Relation.Nullary.Negation.Core using (¬_)
+open import Relation.Binary
+  using (Setoid; IsEquivalence; Reflexive; Symmetric; Transitive)
+open import Relation.Binary.PropositionalEquality
+  using (_≡_; cong₂) renaming (refl to ≡-refl; isEquivalence to ≡-isEquivalence)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+
+
+------------------------------------------------------------------------
+-- 'pre'-free algebra
+
+infixl 7 _∙_
+
+data PreFreeMagma {a} (A : Set a) : Set a where
+
+  var : A → PreFreeMagma A
+  _∙_ : Op₂ (PreFreeMagma A)
+
+module _ {a b} {A : Set a} {B : Set b} where
+
+  map : (A → B) → PreFreeMagma A → PreFreeMagma B
+  map f (var a) = var (f a)
+  map f (s ∙ t) = (map f s) ∙ (map f t)
+
+module _ {a b c} {A : Set a} {B : Set b} {C : Set c} where
+
+  map-id : ∀ (t : PreFreeMagma A) → map id t ≡ t
+  map-id (var a) = ≡-refl
+  map-id (s ∙ t) = cong₂ _∙_ (map-id s) (map-id t)
+
+  map-∘ : (g : A → B) → (f : B → C) → ∀ t → map (f ∘ g) t ≡ (map f ∘ map g) t
+  map-∘ g f (var a) = ≡-refl
+  map-∘ g f (s ∙ t) = cong₂ _∙_ (map-∘ g f s) (map-∘ g f t)
+
+------------------------------------------------------------------------
+-- Functor, RawMonad instance: TODO
+
+------------------------------------------------------------------------
+-- parametrised 'equational' theory over the 'pre'-free algebra
+
+module PreFreeTheory {c ℓ} {A : Set c} (R : Rel A ℓ) where
+
+  data PreFreeMagmaTheory : Rel (PreFreeMagma A) (c ⊔ ℓ)
+
+  open Defs PreFreeMagmaTheory
+
+  data PreFreeMagmaTheory where
+
+    var : ∀ {a b} → (R a b) → PreFreeMagmaTheory (var a) (var b)
+    _∙_ : Congruent₂ _∙_
+
+PreFreeTheorySyntax : ∀ {c ℓ} {A : Set c} (R : Rel A ℓ) → Rel (PreFreeMagma A) (c ⊔ ℓ)
+PreFreeTheorySyntax R = PreFreeMagmaTheory where open PreFreeTheory R
+
+syntax PreFreeTheorySyntax R m n = m ≈[ R ] n
+
+
+------------------------------------------------------------------------
+-- Free algebra on a Set
+{-
+   in the propositional case, we can immediately define the following
+   but how best to organise this under the Algebra.Bundles.Free hierarchy?
+   e.g. should we distinguish Free.Magma.Setoid from Free.Magma.Propositional?
```

Comment:
Unfortunately, I think the `Setoid` and propositional cases do need to be separated. Reason: in Agda, some free algebras don't exist in the propositional case. Example: free Commutative Monoid exists over `Setoid` but not over `Set`.

Reply:
Good point. I think you will have detected my reticence about `Set` structure only being what I dubbed 'pre-free', because raw syntax doesn't admit the right equations once the algebra gets interesting. I might now call this `RawFree` or `FreeRaw`. But in the case, it ok because `Magma` had no equations?

Reply:
It's ok for all theories that have no equations; you can have some equations too, but these have to be 'linear' and 'rigid' - but that's another matter. So it will work for all theories X where the term algebra is the free X for sure.